datadog/resource_datadog_role.go

```diff
@@ -119,102 +120,151 @@ func resourceDatadogRoleCreate(d *schema.ResourceData, meta interface{}) error {
 	roleData := resp.GetData()
 	d.SetId(roleData.GetId())
 
-	return resourceDatadogRoleRead(d, meta)
+	return updateRoleState(d, roleData.Attributes, roleData.Relationships, client, auth)
 }
 
-func resourceDatadogRoleRead(d *schema.ResourceData, meta interface{}) error {
-	client := meta.(*ProviderConfiguration).DatadogClientV2
-	auth := meta.(*ProviderConfiguration).AuthV2
-
-	// Get the role
-	resp, httpresp, err := client.RolesApi.GetRole(auth, d.Id()).Execute()
-	if err != nil {
-		if httpresp != nil && httpresp.StatusCode == 404 {
-			d.SetId("")
-			return nil
-		}
-		return translateClientError(err, "error getting role")
+func updateRoleState(d *schema.ResourceData, roleAttrsI interface{}, roleRelations *datadog.RoleResponseRelationships, client *datadog.APIClient, auth context.Context) error {
+	type namer interface {
+		GetName() string
 	}
-	roleData := resp.GetData()
-	roleAttrs := roleData.GetAttributes()
-	if err := d.Set("user_count", roleAttrs.GetUserCount()); err != nil {
-		return err
-	}
-	if err := d.Set("name", roleAttrs.GetName()); err != nil {
-		return err
+	if roleAttrsI != nil {
+		switch roleAttrs := roleAttrsI.(type) {
+		case *datadog.RoleAttributes:
+			if err := d.Set("user_count", roleAttrs.GetUserCount()); err != nil {
+				return err
+			}
+			if err := d.Set("name", roleAttrs.GetName()); err != nil {
+				return err
+			}
+		case *datadog.RoleUpdateAttributes, *datadog.RoleCreateAttributes:
+			if err := d.Set("name", roleAttrs.(namer).GetName()); err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unexpected type %s for role attributes", reflect.TypeOf(roleAttrsI).String())
+		}
 	}
 
+	rolePerms := roleRelations.GetPermissions()
+	return updateRolePermissionsState(d, rolePerms.GetData(), client, auth)
+}
+
+func updateRolePermissionsState(d *schema.ResourceData, rolePermsI interface{}, client *datadog.APIClient, auth context.Context) error {
+
 	// Get a list of all valid permissions, to ignore restricted perms
 	permsIDToName, err := getValidPermissions(client, auth)
 	if err != nil {
 		return err
 	}
 
-	roleRelations := roleData.GetRelationships()
-	rolePerms := roleRelations.GetPermissions()
-	rolePermsData := rolePerms.GetData()
 	var perms []map[string]string
-	for _, perm := range rolePermsData {
-		permID := perm.GetId()
-		// If perm ID is not restricted, add it to the state
-		if permName, ok := permsIDToName[permID]; ok {
-			permR := map[string]string{
-				"id":   permID,
-				"name": permName,
-			}
-			perms = append(perms, permR)
+	switch rolePerms := rolePermsI.(type) {
+	case []datadog.RelationshipToPermissionData:
+		for _, permI := range rolePerms {
+			perms = appendPerm(perms, &permI, permsIDToName)
```

Comment:
I think you can just pass `permI.GetId()`, this will simplify the code in `appendPerm`.

Reply:
True, thanks for the suggestion